### PR TITLE
Miscellaneous improvements

### LINKED ISSUE #320: Unable to wrap `stat`
This happened when I am wrapping GDAL.jl.
On linux, if I attemp to wrap this code:
```c
#include <sys/stat.h>
typedef struct stat mystat;
```
because `stat` is also a function, `Audit` pass can not be passed. But it seems that neither `CollectTopLevelNode` nor `CollectDependentSystemNode` supports filtering, and `Audit` pass can not be disabled.
Do we need to add filtering earlier to solve this?

### LINKED ISSUE #325: Wrong order of generated structs for generation of structs containing anonymous unions
Hi everyone, 
thank you for the amazing package so far!

I think I encountered a small (but annoying) bug in the current master branch: the generation of structs containing anonymous unions prints the corresponding Julia union wrapper struct after the original struct definition. 

In my case, the C struct

```C  
typedef struct {
      enum NodeIdType id_type;
      union {
          UInt32     numeric;
          Guid       guid;
      } id;
  } NodeId;
```
results in the generated code

```julia
mutable struct NodeId
    id_type::NodeIdType
    id::__JL_Ctag_295
    NodeId() = new()
end
...
struct __JL_Ctag_295
    data::NTuple{16, UInt8}
end
```
leading to 
```
ERROR: LoadError: LoadError: UndefVarError: __JL_Ctag_295 not defined
```
If I am not missing something, the problem can be solved by changing the order of the generated structs.

### LINKED ISSUE #326: Can not handle pointer to anonymous struct
```c
typedef struct {
    int a, b;
} B, *C;
```
Clang.jl can not distinguish B from C.

### PR DESCRIPTION
**Pull request by melonedo**
- [x] Improve documentations
- [x] Support compilation database
- [x] Update FreeBSD version from 11 to 12 (JuliaPackaging/BinaryBuilderBase.jl#166)
- [x] Support seperating common and API files 
- [x] Support topological sorting for macros 

Fix #320, fix #325, fix #326.

### CODE REVIEW

**Comment by melonedo** [comment]
It turns out that it is very easy to split common file and API file and topo-sort the macros. This makes resurrecting old scripts much easier.

**Comment by melonedo** [comment]
> what's the use case of caching these top-level decls.

I needed a way to know what unnamed tags refer to, so I cached them. On other passes this is handled by "nested_tags" option, which I were not aware of when implementing this.

**Review by Gnimuc — COMMENTED** [review]
As for #325, I think it can be fixed by merely caching those inner nested tag types. The line needed to be fixed is
https://github.com/JuliaInterop/Clang.jl/blob/50670a80c38bb73e0772debf16fe68040a4654e9/src/generator/resolve_deps.jl#L154-L156

File: src/generator/print.jl [hunk 1]
```diff
@@ -110,6 +110,7 @@ function pretty_print(io, node::ExprNode{<:AbstractStructNodeType}, options::Dic
     inline = struct_field_comment_style == "inline"
     struct_def = node.exprs[1]
     mutable, name, members = struct_def.args
+    name = Base.sym_to_string(name)
```
[the remark this answers is not in the document]
Comment (melonedo) [reply src/generator/print.jl:113]: with `use_deterministic_symbol` set to false,
```c
struct S {
    struct {
        int a;
    } b;
};
```
output would have been
```julia
mutable struct ##Ctag#499
    a::Cint
    var"##Ctag#499"() = new()
end
```

**Comment by Gnimuc** [comment]
> By the way, why nested_tags is index by tag instead of by cursor? The latter feels more natural for me

Because `CLCursor`s are not unique across different translation units.

**Comment by melonedo** [comment]
> > By the way, why nested_tags is index by tag instead of by cursor? The latter feels more natural for me
> 
> Because `CLCursor`s are not unique across different translation units.

Thanks! I now see that they are compared by actual declaration locations.

**Comment by Gnimuc** [comment]
The ultimate solution would be to directly work with Clang's `ASTContext` and do AST-merging for CTU analysis. Due to libclang's limitations, we have to use these ugly workarounds for now.

**Comment by Gnimuc** [comment]
Is this ready for review? better to split this to individual PRs.

**Comment by melonedo** [comment]
Yes this is ready, this branch was meant to hold commits that might depend on each other, which now seems not happening a lot. I will split them into individual PRs soon.

**Comment by melonedo** [comment]
All of the commits are moved to other branches. Closed.

**Pull request closed without merging**